Widgets: Porting of WP.com's EU Cookie Law Banner

Copons opened the pull request:
This is my first PR in the effort of filling the gaps between Jetpack and WP.com, to smooth the AT process out.

I'm not completely sure about the Jetpack coding standards (file, classes and variables naming conventions, documentation) as apparently different widgets have different standards, so if you have any suggestions I'll be happy to implement them ASAP.

#### Changes proposed in this Pull Request:

* Adds the EU Cookie Law Banner widget as already available on WP.com.

#### Testing instructions:

* Add the widget to a sidebar and check if the banner shows up near the bottom of the window.
* Check if, when closed with the selected method, it actually stays closed.
* Check if both light and dark themes work appropriately.

zinigor commented on modules/widgets/eu-cookie-law.php:
```diff
@@ -0,0 +1,164 @@
+
+		do_action( 'jetpack_stats_extra', 'widget_view', 'eu_cookie_law' );
+
+		echo 'CIAO';
```
I'm pretty sure that's not supposed to be here :)

Copons replied:
That's totally required by EU laws!

Copons commented:
@zinigor thanks for giving a look!
I've labeled In Progress mostly because I wanted more time to look at documentations standards in other JP widgets.
I could have reused this widget .com version docblocks, but there are internal links that I don't think are supposed to get into JP.

Stoyan0v commented on modules/widgets/eu-cookie-law/footer.php:
```diff
@@ -0,0 +1,77 @@
+		_e( 'To find out more, as well as how to remove or block these, see here:', 'jetpack' );
+		echo ' ';
+	} else {
+		echo esc_html( $instance['customtext'] ), ' ';
```
One thing that I notice is that you should put `,'  '`  in the brackets, and you should replace the empty space with jetpack textdomain. 
I started this this widget while ago, but I hadn't enough time to complete it. Feel free to use any codeblock you like https://github.com/Automattic/jetpack/compare/master...Stoyan0v:add/eu-cookie-law-widget

Stoyan0v commented on modules/widgets/eu-cookie-law/footer.php:
```diff
@@ -0,0 +1,77 @@
+		echo esc_html( $instance['customtext'] ), ' ';
+	} ?>
+
+	<a href="<?php echo esc_attr(
```
I think that using [esc_url](https://codex.wordpress.org/Function_Reference/esc_url) is better option here.
And Also it will be nice if you move the logic outside of function.

jeherve commented on modules/widgets/eu-cookie-law.php:
```diff
@@ -0,0 +1,162 @@
+	function __construct() {
+		parent::__construct(
+			'eu_cookie_law_widget',
+			apply_filters( 'jetpack_widget_name', esc_html__( 'EU Cookie Law Banner', 'jetpack' ) ),
```
All code in Jetpack is parsed after each release to build our codex:
http://developer.jetpack.com/

The codex references all actions and hooks in Jetpack, in order to help third-party developers who would like to customize Jetpack.

That means that you'll need to add docblocks for each one of the hooks you add in Jetpack. In this specific case, since the filter isn't new, this should do the trick:
`/** This filter is documented in modules/widgets/facebook-likebox.php */`

jeherve commented on modules/widgets/eu-cookie-law.php:
```diff
@@ -0,0 +1,162 @@
+			'customtext'         => '',
+			'color-scheme'       => 'default',
+			'policy-url'         => 'default',
+			'default-policy-url' => 'https://en.support.wordpress.com/cookies',
```
Since Jetpack and WordPress.com use different cookies, you'll probably need to create a new support doc on jetpack.com for this widget.

jeherve commented on modules/widgets/eu-cookie-law.php:
```diff
@@ -0,0 +1,162 @@
+			'custom-policy-url'  => '',
+			'policy-link-text'   => __( 'Our Cookie Policy', 'jetpack' ),
+			'button'             => __( 'Close and accept', 'jetpack' ),
+			'default-text'       => __( 'Privacy & Cookies: This site uses cookies from WordPress.com and selected partners. ', 'jetpack' ),
```
Since this site is self-hosted, it doesn't use cookies from WordPress.com (at least not for logged out readers).

jeherve commented on modules/widgets/eu-cookie-law.php:
```diff
@@ -0,0 +1,162 @@
+	public function widget( $args, $instance ) {
+		$this->instance = wp_parse_args( $instance, $this->defaults );
+
+		do_action( 'jetpack_stats_extra', 'widget_view', 'eu_cookie_law' );
```
`/** This action is already documented in modules/widgets/gravatar-profile.php */`

jeherve commented on modules/widgets/eu-cookie-law.php:
```diff
@@ -0,0 +1,162 @@
+		require( dirname( __FILE__ ) . '/eu-cookie-law/footer.php' );
+	}
+
+	public function form( $instance ) {
```
It would be a nice plus to add docblocks before each one of your functions. If you install `phpcs`, you'll see that it complains when no comments are added:
https://github.com/Automattic/jetpack/blob/master/docs/coding-guidelines.md#general

Copons replied:
I had to disable phpcs since it complained about the name of every class file. 😄

jeherve commented on modules/widgets/eu-cookie-law.php:
```diff
@@ -0,0 +1,162 @@
+	}
+
+	function enqueue_style() {
+		wp_enqueue_style( 'eu-cookie-law-style', plugins_url( 'eu-cookie-law/style.css', __FILE__ ), array(), '20170403' );
```
Jetpack automatically concatenates all CSS files into a single `jetpack.css` file. To add your CSS to the concatenated file, you'll need to add it [here](https://github.com/Automattic/jetpack/blob/master/gulpfile.js#L291) and [here](https://github.com/Automattic/jetpack/blob/master/class.jetpack.php#L58).

jeherve commented on modules/widgets/eu-cookie-law/footer.php:
```diff
@@ -0,0 +1,77 @@
+</div>
+
+<script type="text/javascript">
+	jQuery( function( $ ) {
```
We rely on jQuery here, but we didn't make sure it was actually enqueued on the page before.

Copons replied:
I believe I should just enqueue this script and pass the values I need from PHP (cookie name and validity) in data attributes instead.

jeherve commented on modules/widgets/eu-cookie-law/form.php:
```diff
@@ -0,0 +1,195 @@
+
+<p class="small">
+	<?php _e( 'It is your own responsibility to ensure that your site complies with the relevant laws.', 'jetpack' ); ?>
+	<a href="">
```
Should there be a link here?

eliorivero commented on modules/widgets/eu-cookie-law/footer.php:
```diff
@@ -0,0 +1,77 @@
+			}
+
+			var expireTime = new Date();
+			expireTime.setTime( expireTime.getTime() + <?php echo $cookie_validity * 1000; ?> );
```
Needs `esc_js`

eliorivero commented on modules/widgets/eu-cookie-law/footer.php:
```diff
@@ -0,0 +1,77 @@
+			var expireTime = new Date();
+			expireTime.setTime( expireTime.getTime() + <?php echo $cookie_validity * 1000; ?> );
+
+			document.cookie = '<?php echo $cookie_name; ?>=' + expireTime.getTime() + ';path=/;expires=' + expireTime.toGMTString();
```
Needs `esc_js`

eliorivero commented on modules/widgets/eu-cookie-law/form.php:
```diff
@@ -0,0 +1,195 @@
+<p>
+	<strong>
+		<?php echo _x( 'Hide the banner', 'action', 'jetpack' ); ?>
```
Could use `_ex`

Copons commented:
@Stoyan0v, @jeherve Thanks so much for the pointers and sorry for the missing stuff: I'm not used to work in PHP anymore and I had quite a long time of trial-and-errors debugging this!

eliorivero commented on modules/widgets/eu-cookie-law/form.php:
```diff
@@ -0,0 +1,195 @@
+	</ul>
+	<textarea
+		class="widefat"
+		name="<?php echo $this->get_field_name( 'text' ); ?>"
```
Needs `esc_attr` like it's done right below in the `placeholder` attribute.

eliorivero commented on modules/widgets/eu-cookie-law/form.php:
```diff
@@ -0,0 +1,195 @@
+			<label>
+				<input
+					<?php checked( $instance['hide'], 'button' ); ?>
+					name="<?php echo $this->get_field_name( 'hide' ); ?>"
```
This and other similar cases need `esc_attr`.

eliorivero commented:
One more:

<img width="268" alt="captura de pantalla 2017-04-04 a las 14 20 44" src="https://cloud.githubusercontent.com/assets/1041600/24669758/feb9f806-1941-11e7-903f-2e764922a9a4.png">

Copons commented:
Thanks @eliorivero!

I didn't expect such a thorough review on an in-progress PR.
You JP guys are so much nicer than those Calypso peeps! 😄

Copons commented:
@eliorivero The various bugs on custom texts should be fixed but I cannot reproduce the dark color scheme button:

<img width="744" alt="screen shot 2017-04-05 at 16 28 34" src="https://cloud.githubusercontent.com/assets/2070010/24713274/16ae35ae-1a1d-11e7-900c-83f9c5d9e436.png">
<img width="744" alt="screen shot 2017-04-05 at 16 28 47" src="https://cloud.githubusercontent.com/assets/2070010/24713275/16c9d700-1a1d-11e7-8128-87e5490cd3ce.png">

May I ask you what browser/OS you're using?

Copons commented:
@jeherve I've discovered that there already is a [Cookies policy page in JP](https://jetpack.com/support/cookies/), so I've simply replaced the old link with this one.

Related to this, I've removed the WP.com mention in the default banner text. Unfortunately, it becomes very short and ugly, but that's an easily fixable copy issue.
About copy, I've added a support page draft for this widget. I'll see to ping someone for the necessary checks.

eliorivero commented:
@Copons looking great!

Sorry I forgot to mention this the first time: I got the white text on light grey button using TwentyTwelve theme. Chrome on El Capitan.

Copons commented:
@eliorivero Thanks! Going to look into it.

Copons commented:
@eliorivero It was a problem with Twenty Twelve styles that used `background-image` for the input background, that overrode this widget's `background-color`.

sirbrillig commented:
This works for me! 

However, I was unable to get the banner to show up in Twenty Seventeen if added to the "Blog Sidebar". It works if added to one of the "Footer" sidebars, though. 

Any sidebar seems to work fine for me in Twenty Fifteen and Twenty Fourteen.

In Twenty Thirteen, it does not hover if not in the "Footer" sidebar:
<img width="750" alt="screen shot 2017-04-20 at 5 12 36 pm" src="https://cloud.githubusercontent.com/assets/2036909/25252850/9335cb52-25ec-11e7-8752-fa7bc608e320.png">

Probably this behavior is expected since I get the same thing on WP.com sites, but I figured I'd mention it for the record since the testing instructions just read "add the widget to a sidebar", and not specifically a "footer" sidebar.

Copons commented:
@sirbrillig At some point I got the same issue you're mentioning here, don't remember on what theme but likely on Twenty Seventeen, so this might be the solution for your problem.
Fact is, the homepage (set as a static page) has no sidebar, so no sidebar widgets will be rendered. If you try going in a blog page, you should see the banner.
It works in the footer because the homepage (every page actually) has the footer widget area. 😄 

I gave it a quick look at Twenty Thirteen and I'd say that the problem there is that the footer area (called "Main Widget Area") is laid out via Masonry that adds an inline `position: absolute` that overrides the `position: fixed` from the widget stylesheet.
I think that all the solutions are just too hacky though: `!important`? Inline `position: fixed`? Ugh!

sirbrillig commented:
oh! You're right about the home page sidebar! Good thinking.

As for Twenty Thirteen, yeah, I don't have any good ideas. In any case it's the same as dotcom.

kwight commented:
@zinigor @jeherve @eliorivero anything else needed here? We're hoping to get it in for deadline next Tues, thanks!

jeherve commented:
@kwight If you've reviewed and tested it, and are happy with it, do not hesitate to approve the PR. Someone else from Pit Crew will come along and test it again as well so it can be merged.

dereksmart commented on modules/widgets/eu-cookie-law/form.php:
```diff
@@ -0,0 +1,195 @@
+
+<p class="small">
+	<?php esc_html_e( 'It is your own responsibility to ensure that your site complies with the relevant laws.', 'jetpack' ); ?>
+	<a href="">
```
No link?

Copons commented:
@dereksmart Damn! I totally forgot about it.

On .com this is a link to a [support page](https://en.support.wordpress.com/cookie-widget).
I've already created a draft on the JP support section (the most recent one), but I'd love a second look by some English speaker.
I'll add the permalink in advance though.

eliorivero commented:
Works fine!
